src/Workspaces/Remote/ServiceHub/Host/SolutionAssetCache.cs

```diff
@@ -11,230 +11,263 @@
-            }
+            // Get the checksums for the local solution.  Note that this will ensure that all child checksums are
+            // computed.  As such, we can just use TryGetXXX below to get them.
+            var compilationState = pinnedSolution.CompilationState;
+            var compilationStateChecksums = await compilationState.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
+            compilationStateChecksums.AddAllTo(pinnedChecksums);
 
-            using (Logger.LogBlock(FunctionId.AssetStorage_CleanAssets, cancellationToken))
+            var solutionState = compilationState.SolutionState;
+            Contract.ThrowIfFalse(solutionState.TryGetStateChecksums(out var stateChecksums));
+            stateChecksums.AddAllTo(pinnedChecksums);
+
+            foreach (var (_, projectState) in solutionState.ProjectStates)
             {
-                // Ensure that if our remote workspace has a current solution, that we don't purge any items associated
-                // with that solution.
-                PooledHashSet<Checksum>? pinnedChecksums = null;
-                try
-                {
-                    foreach (var (checksum, entry) in _assets)
-                    {
-                        // If not enough time has passed, keep in the cache.
-                        if (entry.Stopwatch.Elapsed <= _purgeAfterTimeSpan)
-                            continue;
-
-                        // If this is a checksum we want to pin, do not remove it.
-                        if (pinnedChecksums == null)
-                        {
-                            pinnedChecksums = PooledHashSet<Checksum>.GetInstance();
-                            await AddPinnedChecksumsAsync(pinnedChecksums, cancellationToken).ConfigureAwait(false);
-                        }
-
-                        if (pinnedChecksums.Contains(checksum))
-                            continue;
-
-                        _assets.TryRemove(checksum, out _);
-                    }
-                }
-                finally
-                {
-                    pinnedChecksums?.Free();
-                }
+                Contract.ThrowIfFalse(projectState.TryGetStateChecksums(out var projectStateChecksums));
+                projectStateChecksums.AddAllTo(pinnedChecksums);
+
+                AddAll(projectState.DocumentStates);
+                AddAll(projectState.AdditionalDocumentStates);
+                AddAll(projectState.AnalyzerConfigDocumentStates);
             }
         }
 
-        private async ValueTask AddPinnedChecksumsAsync(HashSet<Checksum> pinnedChecksums, CancellationToken cancellationToken)
+        void AddAll<TState>(TextDocumentStates<TState> states) where TState : TextDocumentState
         {
-            if (_remoteWorkspace is null)
-                return;
-
-            var checksums = await _remoteWorkspace.CurrentSolution.CompilationState.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
the old logic was virtually useless.  all we pinned were the top solution-compilation-state checksums.  Which basically is almost nothing.  And it certainly doesn't pin *useful* data, like the text documents we are looking at in a particular solution.    I was able to observe this being a problem when long running operations were going on such that the operation went past the 'purge' time for this cache.  the cache would drop important data (like documents), despite being referenced by an in-flight solution.  Then, later oop requests would have to send that same information over again.